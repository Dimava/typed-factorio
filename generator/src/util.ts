export function sortByOrder(a: { order: number }, b: { order: number }): number {
  return a.order - b.order
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function assertNever(_value: never): never {
  throw new Error("Value should be never")
}

export function getFirst<T>(elemOrArray: T | T[]): T {
  if (Array.isArray(elemOrArray)) return elemOrArray[0]
  return elemOrArray
}
