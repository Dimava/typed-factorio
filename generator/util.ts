export function byOrder(a: { order: number }, b: { order: number }): number {
  return a.order - b.order
}

export function assertNever(value: never): never {
  throw new Error("Value should be never: " + value)
}

export function getFirst<T>(elemOrArray: T | T[]): T {
  if (Array.isArray(elemOrArray)) return elemOrArray[0]
  return elemOrArray
}
