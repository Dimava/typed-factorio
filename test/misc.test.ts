import { tstl } from "./util"

test("hello world", () => {
  tstl`
  function greetPlayer(player: LuaPlayer) {
    player.print(serpent.line({ hello: "world!" }))
    player.print(true)
    player.print(undefined)
  }

  script.on_event(defines.events.on_player_created, (e) => {
    greetPlayer(game.get_player(e.player_index)!)
  })

  script.on_init(() => {
    for (const [, luaPlayer] of pairs(game.players)) {
      greetPlayer(luaPlayer)
    }
  })
  `
})

test("Lua lazy loaded value", () => {
  tstl`
  script.on_event(defines.events.on_player_setup_blueprint, (e) => {
    const mapping: Record<number, LuaEntity> = e.mapping.get()
  })
  `.expectToHaveNoDiagnostics()
})

test("array-like", () => {
  tstl`
  const myInventory = game.create_inventory(4)

  const slot1 = myInventory[0]
  const slot2 = myInventory[1]
  const len = myInventory.length
  `.expectToMatchSnapshot()
})

test("pairs iteration on lua custom tables", () => {
  tstl`
  for (const [id, player] of pairs(game.players)) {
    const num: number = id
    const p: LuaPlayer = player
  }
  const bar: Record<string, object> = { foo: {} }

  for (const [key, value] of pairs(bar)) {
    const num: string = key
    const p: object = value
  }
  `.expectToHaveNoDiagnostics()
})
