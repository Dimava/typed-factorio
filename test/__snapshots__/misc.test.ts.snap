// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array-like 1`] = `
"myInventory = game.create_inventory(4)
slot1 = myInventory[1]
slot2 = myInventory[2]
len = #myInventory
"
`;
